### ISSUE TITLE
Customize claims in refresh tokens

### ISSUE THREAD

**Issue by JauernigIT** [openiddict/openiddict-core#1330]
I noticed that the refresh tokens are (often unnecessarily) long, because they contain all claims and you cannot configure which claims should be copied to it (by the claim destination).

I suggest to make the refresh token configurable in the claim destinations as well. Or is there a good reason to hold all claims in the refresh token?

**Comment by kinosang** [comment]
`UseReferenceRefreshTokens`

**Comment by JauernigIT** [comment]
As I understood, `UseReferenceRefreshTokens` will store the claims/payload in the database. This is not what I have intended: to have only necessary claims in the refresh token and not all claims that are also stored in the access/id token.

**Comment by JauernigIT** [comment]
My best solution so far is to add an event handler and just keep the necessary claims manually:

````csharp
// modify the encrypted refresh token to only include necessary claims
options.AddEventHandler<ProcessSignInContext>(builder =>
{
    builder.SetOrder(GenerateIdentityModelRefreshToken.Descriptor.Order - 1)
        .AddFilter<RequireRefreshTokenGenerated>()
        .SetType(OpenIddict.Server.OpenIddictServerHandlerType.Custom)
        .UseInlineHandler(context =>
        {
            context.RefreshTokenPrincipal = context.RefreshTokenPrincipal.Clone(
                claim => claim.Type is (
                    Claims.Private.AuthorizationId or
                    Claims.Private.Presenter or
                    Claims.Private.TokenId or
                    Claims.Private.Scope or
                    Claims.Subject or
                    Claims.ExpiresAt
                )
            );
            return default;
        });
});
````

**Comment by kevinchalet** [comment]
> Or is there a good reason to hold all claims in the refresh token?

It's to allow implementations that can't re-hydrate user claims to reuse the claims added when handling the initial authorization request (e.g when using Integrated Windows Authentication or an external authentication provider without a backing user database).

> I suggest to make the refresh token configurable in the claim destinations as well.

When I added the claims destinations feature, I decided against this for two reasons:
  - This feature is mainly a security feature meant to ensure sensitive claims don't end up being leaked (identity tokens are typically not encrypted and access tokens may or may not be encrypted). It's not been designed to make authorization codes or refresh tokens lighter.
  - It was hard for some people to realize that they had to explicitly add "code" and "refresh_token" to all their access/identity token claims for the authorization code flow and refresh token flow to work correctly: without these 2 destinations, the claims would be simply lost during the 2-phase authorization flow.

I'm not against revisiting this design, but since it will be a major breaking change, it will be hard to sell.

That said, this scenario is very easy to implement using the events model.
